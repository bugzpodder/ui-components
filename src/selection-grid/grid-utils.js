// @flow
import { generateFilledArray, getColHeader, getRowHeader } from "@grailbio/lib";

export const validateGridData = (numRows: number, numCols: number, gridData: Array<Array<any>>) => {
  if (numRows !== gridData.length) {
    throw new Error(`Number of rows in gridData (${gridData.length}) does not match numRows (${numRows})`);
  }
  gridData.forEach(gridRow => {
    if (gridRow.length !== numCols) {
      throw new Error(`Number of columns in gridData (${gridRow.length}) does not match numCols (${numCols})`);
    }
  });
  return true;
};

export const generateColHeaders = (numCols: number) => generateFilledArray(numCols, (_, index) => getColHeader(index));

export const generateRowHeaders = (numRows: number) => generateFilledArray(numRows, (_, index) => getRowHeader(index));
