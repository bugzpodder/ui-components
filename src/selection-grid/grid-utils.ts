import { generateFilledArray, getColHeader, getRowHeader } from "@grailbio/lib";

export const validateGridData = (
  numRows: number,
  numCols: number,
  gridData: any[][],
): boolean => {
  if (numRows !== gridData.length) {
    throw new Error(
      `Number of rows in gridData (${gridData.length}) does not match numRows (${numRows})`,
    );
  }
  gridData.forEach(gridRow => {
    if (gridRow.length !== numCols) {
      throw new Error(
        `Number of columns in gridData (${gridRow.length}) does not match numCols (${numCols})`,
      );
    }
  });
  return true;
};

export const generateColHeaders = (numCols: number): string[] =>
  generateFilledArray(numCols, (_, index) => getColHeader(index));

export const generateRowHeaders = (numRows: number): string[] =>
  generateFilledArray(numRows, (_, index) => getRowHeader(index));
