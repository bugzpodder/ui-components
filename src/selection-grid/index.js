// @flow
export { SelectionGrid } from "./selection-grid";
