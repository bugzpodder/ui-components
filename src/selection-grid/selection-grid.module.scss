@import "../styles/colors.scss";

.grid {
  .row {
    display: flex;
    justify-content: flex-start;
    align-items: center;
    .header,
    .cell {
      flex: 1;
      text-align: center;
    }
    .header {
      display: flex;
      justify-content: center;
      align-items: center;
      color: black;
      min-height: 40px;
    }
    .cell {
      position: relative;
      height: 80px;
      background-color: $grail-purple-dark-color;
      color: white;
      border: 1px solid $grail-gray-light-color;
      &.isEmpty {
        background-color: $grail-gray-lighter-color;
        border: 1px solid white;
      }
      &.isSelected {
        border-color: $grail-tan-color;
      }
      &.isSelectable {
        &:hover {
          cursor: pointer;
        }
      }

      .cellContent {
        font-size: 12px;
        margin: 0;
        height: 100%;
        display: flex;
        justify-content: center;
        align-items: center;
        &.isSelected {
          border: 4px solid $grail-tan-color;
        }
      }
    }
  }
}
