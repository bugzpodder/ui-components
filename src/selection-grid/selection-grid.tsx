import React, { ReactNode } from "react";
import classNames from "classnames";
import styles from "./selection-grid.module.scss";
import { GridCell } from "./components/grid-cell";
import {
  GridCellCoordinate,
  GridCellInfo,
  SelectionGridClasses,
} from "../types/selection-grid";
import { GridHeader } from "./components/grid-header";
import {
  generateColHeaders,
  generateRowHeaders,
  validateGridData,
} from "./grid-utils";

type Props = {
  /** Data to be displayed in grid */
  gridData: any[][];
  /** Number of rows in the grid */
  numRows: number;
  /** Number of columns in the grid */
  numCols: number;
  /** A set of selected cell indices */
  selectedCoordinates: GridCellCoordinate[];
  /** Custom cell rendering function */
  cellRenderer: (x0: GridCellInfo) => ReactNode;
  /** Onclick handler for selecting cell */
  onSelect?: (x0: GridCellInfo) => void;
  /** Dictates whether to show the row/column header */
  showHeader?: boolean;
  /**
   * Provides classNames to the grid and its sub-components. Options include:
   *
   *  - `root`
   *
   *  - `header` (applied to the row/column headers)
   *
   *  - `row` (applied to each row of the grid. Can take a function `(rowIndex: number) => classname: string` to help
   *    specify a className for any specific row)
   *
   *  - `cell` (applied to each cell. Can take a function `(gridCellInfo: GridCellInfo) => classname: string` to help
   *    specify a className for any specific cell)
   *
   *  - `cellContent` (applied to content div inside each cell)
   */
  classes?: SelectionGridClasses;
  /** Should return true if cell is selectable. By default, it'll check if the instance at the position is not null. */
  isCellSelectable?: (x0: GridCellInfo) => boolean;
  /** Should return true if cell is empty. By default, it'll check if the instance at the position is null. */
  isCellEmpty?: (x0: GridCellInfo) => boolean;
  /** Invert rows showing from largest row to smallest */
  invertRows?: boolean;
};

export const SelectionGrid: React.FC<Props> = props => {
  const {
    gridData,
    numCols,
    numRows,
    selectedCoordinates,
    classes = {},
    showHeader = true,
    onSelect,
    cellRenderer,
    isCellSelectable = ({ instance }) => instance != null,
    isCellEmpty = ({ instance }) => instance == null,
    invertRows,
  } = props;

  try {
    validateGridData(numRows, numCols, gridData);
  } catch (error) {
    console.error(error.message);
    throw error;
  }

  const colHeaders = generateColHeaders(numCols);
  const rowHeaders = generateRowHeaders(numRows);
  const calculateCellIndex = ({ rowIndex, colIndex }: GridCellCoordinate) => {
    return rowIndex * numCols + colIndex;
  };

  const selectedIndices = new Set();
  selectedCoordinates.forEach(coordinate => {
    const { rowIndex, colIndex } = coordinate;
    if (
      rowIndex >= 0 &&
      rowIndex < numRows &&
      colIndex >= 0 &&
      colIndex < numCols
    ) {
      selectedIndices.add(calculateCellIndex(coordinate));
    }
  });

  const rows = gridData.map((row, rowIndex) => {
    let rowClass = classes.row;
    if (typeof classes.row === "function") {
      rowClass = classes.row(rowIndex);
    }
    return (
      <div
        data-testid={`selection-grid-row-${rowIndex}`}
        className={classNames(styles.row, rowClass)}
        key={rowIndex}
      >
        {showHeader && (
          <GridHeader classes={classes}>{rowHeaders[rowIndex]}</GridHeader>
        )}
        {row.map((instance, colIndex) => {
          const cellIndex = rowIndex * numCols + colIndex;
          const gridCellInfo = {
            rowIndex,
            colIndex,
            cellIndex,
            instance,
          };
          return (
            <GridCell
              key={cellIndex}
              gridCellInfo={gridCellInfo}
              classes={classes}
              isSelectable={!!onSelect && isCellSelectable(gridCellInfo)}
              isSelected={selectedIndices.has(cellIndex)}
              isEmpty={isCellEmpty(gridCellInfo)}
              onSelect={onSelect}
              cellRenderer={cellRenderer}
            />
          );
        })}
      </div>
    );
  });

  return (
    <div
      data-testid="selection-grid"
      className={classNames(styles.grid, classes.root)}
    >
      {showHeader && (
        <div className={classNames(styles.row, classes.row)}>
          <GridHeader classes={classes} />
          {colHeaders.map((header, index) => {
            return (
              <GridHeader key={index} classes={classes}>
                {header}
              </GridHeader>
            );
          })}
        </div>
      )}
      {invertRows ? rows.reverse() : rows}
    </div>
  );
};
