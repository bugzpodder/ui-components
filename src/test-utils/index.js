// @flow
export {
  ALPHABET_ANIMALS_DATA,
  ADAMS_DATA,
  COUNTRIES,
  EMU_DATA,
  EXTRA_LARGE_EXAMPLE_TABLE_DATA,
  LARGE_EXAMPLE_TABLE_DATA,
  EXAMPLE_TABLE_DATA,
  GRAIL,
} from "./constants";
export { ExampleBlock } from "./example-block";
export { TestWrapper } from "./test-wrapper";
export { StyleWrapper as ExampleWrapper } from "../style-wrapper";
