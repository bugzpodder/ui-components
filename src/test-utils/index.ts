export {
  ALPHABET_ANIMALS_DATA,
  ADAMS_DATA,
  COUNTRIES,
  COUNTRY_NAMES,
  EMU_DATA,
  EXTRA_LARGE_EXAMPLE_TABLE_DATA,
  LARGE_EXAMPLE_TABLE_DATA,
  EXAMPLE_TABLE_DATA,
  GRAIL,
  TEST_EXTERNAL_DOMAINS,
} from "./constants";
export { ExampleBlock } from "./example-block";
export { TestWrapper } from "./test-wrapper";
export { StyleWrapper as ExampleWrapper } from "../style-wrapper";
