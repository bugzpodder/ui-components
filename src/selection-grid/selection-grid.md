### Example

```js
import { ALPHABET_ANIMALS_DATA, ExampleWrapper } from "../test-utils";
import { useState } from "react";
import cloneDeep from "lodash/cloneDeep";
import { SelectionGrid } from "./";

const ExampleSelectionGrid = () => {
  const [selectedCoordinates, setSelectedCoordinates] = useState([]);
  const numRows = ALPHABET_ANIMALS_DATA.length;
  const numCols = ALPHABET_ANIMALS_DATA[0].length;
  const cellRenderer = gridCellInfo => {
    const { rowIndex, colIndex, cellIndex, instance: animal } = gridCellInfo;
    if (isCellEmpty(gridCellInfo)) {
      return null;
    }
    return (
      <div>
        <div>Row index: {rowIndex}</div>
        <div>Col index: {colIndex}</div>
        <div>Cell index: {cellIndex}</div>
        <div>{animal}</div>
      </div>
    );
  };
  const isCellEmpty = ({ instance }) => instance == null;
  const isCellSelectable = ({ instance }) => instance != null;
  const onSelect = ({ rowIndex, colIndex }) => {
    const coordinates = cloneDeep(selectedCoordinates);
    const index = coordinates.findIndex(
      coordinate => coordinate.rowIndex === rowIndex && coordinate.colIndex === colIndex,
    );
    if (index >= 0) {
      coordinates.splice(index, 1);
    } else {
      coordinates.push({ rowIndex, colIndex });
    }
    setSelectedCoordinates(coordinates);
  };

  return (
    <SelectionGrid
      gridData={ALPHABET_ANIMALS_DATA}
      numRows={numRows}
      numCols={numCols}
      selectedCoordinates={selectedCoordinates}
      cellRenderer={cellRenderer}
      isCellEmpty={isCellEmpty}
      isCellSelectable={isCellSelectable}
      onSelect={onSelect}
    />
  );
};

<ExampleWrapper>
  <ExampleSelectionGrid />
</ExampleWrapper>;
```
