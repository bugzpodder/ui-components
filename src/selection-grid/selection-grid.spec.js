// @flow
import "jest-dom/extend-expect";
import React from "react";
import mockConsole from "jest-mock-console";
import { ALPHABET_ANIMALS_DATA, TestWrapper } from "../test-utils";
import { SelectionGrid } from "./selection-grid";
import { cleanup, fireEvent, render } from "react-testing-library";

afterEach(cleanup);

const selectedCoordinates = [];
const numRows = ALPHABET_ANIMALS_DATA.length;
const numCols = ALPHABET_ANIMALS_DATA[0].length;

const isCellEmpty = ({ instance }) => instance == null;
const isCellSelectable = ({ cellIndex }) => cellIndex % 2 === 0;
const cellRenderer = (gridCellInfo) => {
  const {
    rowIndex, colIndex, cellIndex, instance: animal,
  } = gridCellInfo;
  if (isCellEmpty(gridCellInfo)) {
    return null;
  }
  return (
    <div>
      <div>Row index: {rowIndex}</div>
      <div>Col index: {colIndex}</div>
      <div>Cell index: {cellIndex}</div>
      <div>isCellSelectable: {isCellSelectable(gridCellInfo)}</div>
      <div>{animal}</div>
    </div>
  );
};
const onSelect = jest.fn();

test("render selection grid and select a cell", () => {
  const { container, getByTestId } = render(
    <TestWrapper>
      <SelectionGrid
        gridData={ALPHABET_ANIMALS_DATA}
        numRows={numRows}
        numCols={numCols}
        selectedCoordinates={selectedCoordinates}
        cellRenderer={cellRenderer}
        isCellEmpty={isCellEmpty}
        isCellSelectable={isCellSelectable}
        onSelect={onSelect}
      />
    </TestWrapper>,
  );

  fireEvent.click(getByTestId("grid-cell-2"));
  expect(onSelect).toHaveBeenCalled();
  expect(container).toMatchSnapshot();
});

test("render selection grid with mismatching numRow", () => {
  mockConsole();
  expect(() => render(
    <SelectionGrid
      gridData={ALPHABET_ANIMALS_DATA}
      numRows={numRows - 1}
      numCols={numCols}
      selectedCoordinates={selectedCoordinates}
      cellRenderer={cellRenderer}
      isCellEmpty={isCellEmpty}
      isCellSelectable={isCellSelectable}
      onSelect={onSelect}
    />
  )).toThrowError();
  expect(console.error).toHaveBeenCalled();
});

test("render selection grid with mismatching numCols", () => {
  mockConsole();
  expect(() => render(
    <SelectionGrid
      gridData={ALPHABET_ANIMALS_DATA}
      numRows={numRows}
      numCols={numCols - 1}
      selectedCoordinates={selectedCoordinates}
      cellRenderer={cellRenderer}
      isCellEmpty={isCellEmpty}
      isCellSelectable={isCellSelectable}
      onSelect={onSelect}
    />
  )).toThrowError();
  expect(console.error).toHaveBeenCalled();
});
