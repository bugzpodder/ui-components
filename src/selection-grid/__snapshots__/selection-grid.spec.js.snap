// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render selection grid and select a cell 1`] = `
<div>
  <div
    class="grid"
  >
    <div
      class="row"
    >
      <div
        class="header"
      />
      <div
        class="header"
      >
        1
      </div>
      <div
        class="header"
      >
        2
      </div>
      <div
        class="header"
      >
        3
      </div>
      <div
        class="header"
      >
        4
      </div>
      <div
        class="header"
      >
        5
      </div>
      <div
        class="header"
      >
        6
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="header"
      >
        A
      </div>
      <div
        class="cell isSelectable"
        data-testid="grid-cell-0"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              0
            </div>
            <div>
              Col index: 
              0
            </div>
            <div>
              Cell index: 
              0
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              alligator
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell"
        data-testid="grid-cell-1"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              0
            </div>
            <div>
              Col index: 
              1
            </div>
            <div>
              Cell index: 
              1
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              bat
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell isSelectable"
        data-testid="grid-cell-2"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              0
            </div>
            <div>
              Col index: 
              2
            </div>
            <div>
              Cell index: 
              2
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              cat
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell"
        data-testid="grid-cell-3"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              0
            </div>
            <div>
              Col index: 
              3
            </div>
            <div>
              Cell index: 
              3
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              dog
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell isSelectable"
        data-testid="grid-cell-4"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              0
            </div>
            <div>
              Col index: 
              4
            </div>
            <div>
              Cell index: 
              4
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              elephant
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell"
        data-testid="grid-cell-5"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              0
            </div>
            <div>
              Col index: 
              5
            </div>
            <div>
              Cell index: 
              5
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              fox
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="header"
      >
        B
      </div>
      <div
        class="cell isSelectable"
        data-testid="grid-cell-6"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              1
            </div>
            <div>
              Col index: 
              0
            </div>
            <div>
              Cell index: 
              6
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              goat
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell"
        data-testid="grid-cell-7"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              1
            </div>
            <div>
              Col index: 
              1
            </div>
            <div>
              Cell index: 
              7
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              horse
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell isSelectable"
        data-testid="grid-cell-8"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              1
            </div>
            <div>
              Col index: 
              2
            </div>
            <div>
              Cell index: 
              8
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              iguana
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell"
        data-testid="grid-cell-9"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              1
            </div>
            <div>
              Col index: 
              3
            </div>
            <div>
              Cell index: 
              9
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              jaguar
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell isSelectable"
        data-testid="grid-cell-10"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              1
            </div>
            <div>
              Col index: 
              4
            </div>
            <div>
              Cell index: 
              10
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              kangaroo
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell"
        data-testid="grid-cell-11"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              1
            </div>
            <div>
              Col index: 
              5
            </div>
            <div>
              Cell index: 
              11
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              lion
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="header"
      >
        C
      </div>
      <div
        class="cell isSelectable"
        data-testid="grid-cell-12"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              2
            </div>
            <div>
              Col index: 
              0
            </div>
            <div>
              Cell index: 
              12
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              monkey
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell"
        data-testid="grid-cell-13"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              2
            </div>
            <div>
              Col index: 
              1
            </div>
            <div>
              Cell index: 
              13
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              narwhal
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell isSelectable"
        data-testid="grid-cell-14"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              2
            </div>
            <div>
              Col index: 
              2
            </div>
            <div>
              Cell index: 
              14
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              octopus
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell"
        data-testid="grid-cell-15"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              2
            </div>
            <div>
              Col index: 
              3
            </div>
            <div>
              Cell index: 
              15
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              penguin
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell isSelectable"
        data-testid="grid-cell-16"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              2
            </div>
            <div>
              Col index: 
              4
            </div>
            <div>
              Cell index: 
              16
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              quail
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell"
        data-testid="grid-cell-17"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              2
            </div>
            <div>
              Col index: 
              5
            </div>
            <div>
              Cell index: 
              17
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              raccoon
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="header"
      >
        D
      </div>
      <div
        class="cell isSelectable"
        data-testid="grid-cell-18"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              3
            </div>
            <div>
              Col index: 
              0
            </div>
            <div>
              Cell index: 
              18
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              seal
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell"
        data-testid="grid-cell-19"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              3
            </div>
            <div>
              Col index: 
              1
            </div>
            <div>
              Cell index: 
              19
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              tiger
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell isSelectable"
        data-testid="grid-cell-20"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              3
            </div>
            <div>
              Col index: 
              2
            </div>
            <div>
              Cell index: 
              20
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              umbrella bird
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell"
        data-testid="grid-cell-21"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              3
            </div>
            <div>
              Col index: 
              3
            </div>
            <div>
              Cell index: 
              21
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              vulture
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell isSelectable"
        data-testid="grid-cell-22"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              3
            </div>
            <div>
              Col index: 
              4
            </div>
            <div>
              Cell index: 
              22
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              whale
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell"
        data-testid="grid-cell-23"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              3
            </div>
            <div>
              Col index: 
              5
            </div>
            <div>
              Cell index: 
              23
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              x-ray tetra
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="header"
      >
        E
      </div>
      <div
        class="cell isSelectable"
        data-testid="grid-cell-24"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              4
            </div>
            <div>
              Col index: 
              0
            </div>
            <div>
              Cell index: 
              24
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              yak
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell"
        data-testid="grid-cell-25"
      >
        <div
          class="cellContent"
        >
          <div>
            <div>
              Row index: 
              4
            </div>
            <div>
              Col index: 
              1
            </div>
            <div>
              Cell index: 
              25
            </div>
            <div>
              isCellSelectable: 
            </div>
            <div>
              zebra
            </div>
          </div>
        </div>
      </div>
      <div
        class="cell isSelectable isEmpty"
        data-testid="grid-cell-26"
      >
        <div
          class="cellContent"
        />
      </div>
      <div
        class="cell isEmpty"
        data-testid="grid-cell-27"
      >
        <div
          class="cellContent"
        />
      </div>
      <div
        class="cell isSelectable isEmpty"
        data-testid="grid-cell-28"
      >
        <div
          class="cellContent"
        />
      </div>
      <div
        class="cell isEmpty"
        data-testid="grid-cell-29"
      >
        <div
          class="cellContent"
        />
      </div>
    </div>
  </div>
</div>
`;
